From 9253164b06de479dfabd256d0e7eaa84585e49e2 Mon Sep 17 00:00:00 2001
From: Valery Volgutov <valery.volgutov@lge.com>
Date: Wed, 2 Sep 2015 17:36:45 +0300
Subject: [PATCH] Correct pixelRatio for glyphs

:Release Notes:
Pixel ratio value is used for initialization material for glyphs
and if it is less than one (for example 0.9), by casting value to
integer we will get zero for pixelRatio and wrong material. As a result
glyphs are not drawn.

:Detailed Notes:
Device pixel ratio should be float value, but inside following function:
qsg_device_pixel_ratio it casts to integer. In later use returned from
mentioned above function value again casts to float.

Vertex shader for glyphs multiplies vertex coordinate value to
devicePixelRatio, but glyphs vertex coordinates are already
calculated with devicePixelRatio, so this additional multiplication
gives wrong scaling and positioning.

:Testing Performed:
webos-pro/master/886

:QA Notes:

:Issues Addressed:
[PLAT-1749] FPS number is not displaying

Upstream-Status: Submitted [https://codereview.qt-project.org/#/c/125105/]
---
 src/quick/scenegraph/qsgdefaultglyphnode_p.cpp | 4 ++--
 src/quick/scenegraph/shaders/textmask.vert     | 2 +-
 2 files changed, 3 insertions(+), 3 deletions(-)

diff --git a/src/quick/scenegraph/qsgdefaultglyphnode_p.cpp b/src/quick/scenegraph/qsgdefaultglyphnode_p.cpp
index b001899..d7e2113 100644
--- a/src/quick/scenegraph/qsgdefaultglyphnode_p.cpp
+++ b/src/quick/scenegraph/qsgdefaultglyphnode_p.cpp
@@ -71,9 +71,9 @@ static inline QVector4D qsg_premultiply(const QVector4D &c, float globalOpacity)
     return QVector4D(c.x() * o, c.y() * o, c.z() * o, o);
 }
 
-static inline int qsg_device_pixel_ratio(QOpenGLContext *ctx)
+static inline qreal qsg_device_pixel_ratio(QOpenGLContext *ctx)
 {
-    int devicePixelRatio = 1;
+    qreal devicePixelRatio = 1;
     if (ctx->surface()->surfaceClass() == QSurface::Window) {
         QWindow *w = static_cast<QWindow *>(ctx->surface());
         if (QQuickWindow *qw = qobject_cast<QQuickWindow *>(w))
diff --git a/src/quick/scenegraph/shaders/textmask.vert b/src/quick/scenegraph/shaders/textmask.vert
index 4c67827..8815b3b 100644
--- a/src/quick/scenegraph/shaders/textmask.vert
+++ b/src/quick/scenegraph/shaders/textmask.vert
@@ -10,5 +10,5 @@ varying highp vec2 sampleCoord;
 void main()
 {
      sampleCoord = tCoord * textureScale;
-     gl_Position = matrix * floor(vCoord * dpr + 0.5) / dpr;
+     gl_Position = matrix * floor(vCoord + 0.5) / dpr;
 }
